Add Pango Renderer

naveen521kk opened the pull request:
## List of Changes
- Fixes #393 
- Create a new class for adding Pango as a backend rendering. Pretty much the same implementation as `Text`. 
- Add some example Scene.
- Something to note I can't change libraries due to the current configuration in poetry. So the thing which happened is I had to change the minimum python 3.6 ->3.7.(*facepalm*)
- Added some new deps. `pangocffi`, `pangocairocffi` and `cairocffi`. Pycairo can't work with cffi binding so had to choose `cairocffi`.(*facepalm*)

## Motivation
Currently, Manim uses a Toy API from Cairo to render `Text` which caused many problems for users for different languages. Pango seems to handle it pretty well. 

## Testing Status
Tried the example scene on Windows 7 and Windows 10 and worked like a charm. Possibly need to test on Linux and Mac.

## Further Comments
Pretty much for me installing the required files was a pain but I finally found how to do so in windows. Soon I will document it.

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

## TODO
- [x] Write documentation of those classes.
- [x]  Write documentation on how to install those. At least for windows, because other OS's has then inbuild. Possibly would require Github Actions to build those `.dll` files. But I got it running using my emacs `.dll` files.(*facepalm*)
- [x]  Add these to changelogs.

Here is a sample gif.

![manim-pango](https://user-images.githubusercontent.com/49693820/94178308-8c546480-feb8-11ea-9172-1278f2fe6824.gif)

leotrs commented:
Definitely an improvement to use Pango.

However, you say we will have to use cairocffi and pangocffi for text rendering, while we will continue to use pycairo for mobjects, is that correct? That seems weird to me.

huguesdevimeux commented:
It does fix #262 as well, no ?

naveen521kk commented:
> However, you say we will have to use cairocffi and pangocffi for text rendering, while we will continue to use pycairo for mobjects, is that correct? That seems weird to me.

Yes pretty much. Both runs side by side no problem on that. There is no open source wrapper for pango othen cffi ones. Cffi ones requires `cairocffi`.( There is one from pygobject, aka `pangocairo`, but it requires full gtk out of scope Manim.). So the best option was to use `cairocffi` along with `pangocffi` and `pangocairocffi`(Both pangocffi and pangocairocffi are different library but is necessary).

naveen521kk commented:
> It does fix #262 as well, no 

My reply for that would be no. Pango renders text pretty good even right to left ones but doesn't allows us to select a file and load font from that. For that we would need to use freetype, but yes, again no open source python wrapper for that sadly.

naveen521kk commented:
I suggest we would need a new repo for windows binaries (`.dll` ) files to be created. Only then I can make the tests pass 😁.

naveen521kk commented:
This is pretty much done. ;-)
Remaining is documentation about installing this in Windows and MacOS(I though it had pango but not) and fixing the CI.
Btw I would like to transfer the repo https://github.com/naveen521kk/manim-windows to ManimCommunity. It contain the files that need to downloaded by windows users for installing Manim after this PR.

naveen521kk commented:
@ViolaBuddy can you confirm whether the things in #393 is fixed? Use `PangoText` class instead of `Text` ;-)

naveen521kk commented:
Tests fail in windows because Ffmpeg site is down or either moved completely. Need to look for alt. Other PR tests run because of cache but here cache need to worked up again and so the error. Also, if we didn't fix it all windows test will in all PR next month sadly. 
https://github.com/ManimCommunity/manim/runs/1162323681?check_suite_focus=true#step:10:73

behackl commented:
Is there a reason to keep `CairoText` around? If Pango is able to fully replace Cairo for text rendering, we should simply do so.

Personally, I am confused if there are two text classes without a very clear distinction. (Especially given the somewhat cryptic warning displayed when using `Text`.)

naveen521kk commented:
> Is there a reason to keep CairoText around? If Pango is able to fully replace Cairo for text rendering, we should simply do so.

Until bugs are fixed(possibly may occur) in `PangoText` class we can have Cairo also.

leotrs commented:
Let's not get rid of anything until we are entirely sure the new class is stable. We've been having a lot of regression bugs lately. 

@naveen521kk, in the interest of minimizing regressions, can we please add tests to this? @ManimCommunity/tests I know that tests involving latex are awful, but can we find a workaround? What about tests for right-to-left script? Fonts? etc

naveen521kk commented:
I can create some tests, yes, for the right to left scripts and stuff. Comparing SVG files is the best way. (Pretty much it is the same always until we change fonts). But pretty much I need to first fix CI for windows and I don't know why readthedocs build failed.

leotrs commented:
Seems it can't find the new dependency pangocffi?

naveen521kk commented:
> Seems it can't find the new dependency pangocffi?

How to fix it? Also, Review at will.

Pretty much only documentation on installation is remaining.

naveen521kk commented:
I have transferred https://github.com/naveen521kk/manim-windows to https://github.com/ManimCommunity/manim-windows. I really don't know that I have permissions for that. Pretty much if I did anything wrong tell me.

naveen521kk commented:
Added quite some tests. Ready for review. What still I need to do is
- [ ] installtion instructions for MacOS
- [ ] Add this to changelog

naveen521kk commented:
Tests are fixed. This PR is ready for review. Also, I need some help on readthedocs build( I really don't know how to fix it).

behackl commented:
RTD fails with

```
Exception occurred:
  File "/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/471/lib/python3.8/site-packages/pangocffi/__init__.py", line 2, in <module>
    from ._generated.ffi import ffi
ModuleNotFoundError: No module named 'pangocffi._generated'
```

and I don't see where the `_generated` is supposed to come from. Do you know that?

naveen521kk commented:
> I don't see where the _generated is supposed to come from.

I think that is somewhat internal for `pangocffi`. A question, does this work on Linux?

behackl commented:
> > I don't see where the _generated is supposed to come from.
> 
> I think that is somewhat internal for `pangocffi`. Possibly, I will make a issue there. A question, does this work on Linux? 

I have not tried it locally yet, but the RTD build basically runs on a Ubuntu docker image. We might be missing some dependencies, which would be a problem because we cannot install system packages as a part of the docbuild process.

naveen521kk commented:
> We might be missing some dependencies, which would be a problem because we cannot install system packages as a part of the docbuild process.

Pango is something which is available by default for Linux(any distro).

behackl commented:
For some reason, the bindings contained in `pangoffi/_generated` are not generated during the setup of the package. I am trying to do that manually now (https://github.com/ManimCommunity/manim/pull/471/commits/9cdd00ed05d88cf66309a7c3540be287a786e94c) -- and it seems that it worked, now `pangocairocffi` is complaining.

behackl commented:
The documentation builds again.

naveen521kk commented:
Looks like something to report to them.
Do you know why the example I wrote here https://manimce--471.org.readthedocs.build/en/471/reference/manim.mobject.svg.text_mobject.PangoText.html#manim.mobject.svg.text_mobject.PangoText didn't render?

behackl commented:
The warning from the docbuild process is

```
/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/471/lib/python3.8/site-packages/manim/mobject/svg/text_mobject.py:docstring of manim.mobject.svg.text_mobject.PangoText:14: WARNING: Error in "manim" directive:
invalid option block.

.. manim:: PangoRender
    :quality: low
    class PangoRender(Scene):
        def construct(self):
            morning = PangoText("வணக்கம்", font="sans-serif")
            self.play(Write(morning))
            self.wait(2)
```

which hints that you are missing a blank line between `:quality: low` and the start of the code block.

behackl commented:
> Looks like something to report to them.

The weird thing is that with my local installation, the automatic generation during `pip install` worked. I am not sure why it didn't on RTD, might be something weird on their side as well.

naveen521kk commented:
> The weird thing is that with my local installation, the automatic generation during pip install worked. I am not sure why it didn't on RTD, might be something weird on their side as well.

I could make out the same error in Google Colab while working on virtualvenv but not on system install. Could you try the same on virtualvenv and report if it caused error?

naveen521kk commented:
All tests pass now. Just need reviews for this to be merged.

leotrs commented on docs/source/conf.py:
```diff
@@ -19,8 +20,12 @@
-    # we need to add ffmpeg to the path
     site_path = get_python_lib()
+    # bindings for pangocffi, cairocffi, pangocairocffi need to be generated
+    subprocess.run(["python", "pangocffi/ffi_build.py"], cwd=site_path)
+    subprocess.run(["python", "cairocffi/ffi_build.py"], cwd=site_path)
+    subprocess.run(["python", "pangocairocffi/ffi_build.py"], cwd=site_path)
```
@naveen521kk how long does this take? 
@behackl what do we think of this?

naveen521kk replied:
I don't know @behackl had this implementation.

behackl replied:
Yes, guilty as charged. I am not aware of any other way that fixes the documentation build. The required time is neglegible:

```
behackl:site-packages/ $ time python pangocffi/ffi_build.py                    [15:54:31]
python pangocffi/ffi_build.py  0,20s user 0,03s system 99% cpu 0,233 total
behackl:site-packages/ $ time python pangocairocffi/ffi_build.py               [15:54:56]
python pangocairocffi/ffi_build.py  0,23s user 0,05s system 100% cpu 0,274 total
behackl:site-packages/ $ time python cairocffi/ffi_build.py                    [15:55:12]
python cairocffi/ffi_build.py  0,22s user 0,02s system 99% cpu 0,244 total
```

Less than a second in total on my local machine, should not be too far off for RTD.

leotrs replied:
Great, thanks!

leotrs commented on docs/source/installation/mac.rst:
```diff
@@ -15,6 +15,16 @@ To install cairo:
 
    brew install cairo
 
+To install Pango and it dependencies:
```
Thank you SO much for adding all these instructinos.

naveen521kk replied:
It should definitely should be part o this PR. ;-)

leotrs commented on manim/constants.py:
```diff
@@ -29,11 +29,23 @@ class MyText(Text):
 OBLIQUE = "OBLIQUE"
 BOLD = "BOLD"
+# Only for Pango from below
+THIN = "THIN"
+ULTRALIGHT = "ULTRALIGHT"
+LIGHT = "LIGHT"
+SEMILIGHT = "SEMILIGHT"
+BOOK = "BOOK"
+MEDIUM = "MEDIUM"
+SEMIBOLD = "SEMIBOLD"
+ULTRABOLD = "ULTRABOLD"
+HEAVY = "HEAVY"
+ULTRAHEAVY = "ULTRAHEAVY"
```
Comment (non-blocking): In the future, we should turn these into an Enum

leotrs commented on tests/test_pango.py:
```diff
@@ -0,0 +1,124 @@
+
+def test_general_text_svgobject():
+    """Checks number of submobjects generated when directly called using ``SVGMobject``"""
+    text = "hello"
+    size = 1
+    folder = os.path.abspath(os.path.join("media", "texts"))
+    if not os.path.exists(folder):
+        os.makedirs(folder)
+    filename = os.path.join(folder, "hello.svg")
```
Can you please make a note somewhere (perhaps a new issue) to get rid of this as soon as SceneFIleWriter and TexFileWriter are refactored? This shouldn't be necessary...

naveen521kk replied:
@behackl already knows this issue and discussed on discord. Also why #476 is pending.

leotrs replied:
I understand that, but what I mean to say is that the real fix to #476 is to refactor SceneFileWriter. Once that is done, we should change this test. That's all I'm saying.

leotrs commented on tests/test_pango.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def test_rtl_text_text_svgobject():
+    """Checks number of submobjects generated when directly called using ``SVGMobject``"""
```
Issue: this is the same as the previous docstring and doesn't seem to describe the test very well.

leotrs commented on tests/test_pango.py:
```diff
@@ -0,0 +1,124 @@
+
+def test_rtl_text_text_svgobject():
+    """Checks number of submobjects generated when directly called using ``SVGMobject``"""
+    size = 1
+    text = rtl_text
+    folder = os.path.abspath(os.path.join("media", "texts"))
+    if not os.path.exists(folder):
+        os.makedirs(folder)
+    filename = os.path.join(folder, "hello.svg")
+    a = PangoText(text, size=1)
+    pangoManim = os.path.join(folder, a.text2hash() + ".svg")
+    surface = cairocffi.SVGSurface(filename, width, height)
+    context = cairocffi.Context(surface)
+    context.move_to(START_X, START_Y)
+    layout = pangocairocffi.create_layout(context)
+    layout.set_width(pangocffi.units_from_double(width))
+    fontdesc = pangocffi.FontDescription()
+    fontdesc.set_size(pangocffi.units_from_double(size * 10))
+    layout.set_font_description(fontdesc)
+    layout.set_text(text)
+    pangocairocffi.show_layout(context, layout)
+    surface.finish()
```
Suggestion: maybe convert this into an auxiliary function? This is all duplicated from the previous test.

leotrs commented on tests/test_pango.py:
```diff
@@ -0,0 +1,124 @@
+    pangocairocffi.show_layout(context, layout)
+    surface.finish()
+    b = SVGMobject(filename)
+    assert len(a.submobjects) == len(b.submobjects)
```
Question: How does this test the font face?

naveen521kk replied:
AFAIK, if Pango uses the default font face for Arabic texts the SVG would not be formatted properly. SO checking for submobject between the one generated directly from Pango and also from manim would do and is a very basic simple check. Also, what are other properties of `mobjects` that can be compared?

leotrs replied:
Not sure I follow. What I'm saying is that the two texts may have the same length of submobjects but have different font face. I'm not sure how to compare the font faces...

naveen521kk replied:
Here you go how Pango has if the font is not there.
![random](https://user-images.githubusercontent.com/49693820/94579494-85519b80-0296-11eb-9a86-71f89c639ab0.png)
While if we select those fonts it will have the same length of submobject(manim parses it so). I also don't know how to check this more exactly.

leotrs replied:
Yeah I see. So this test is not really checking for the font. In that case, we should remove it.

One way to go about it is to use the same framework as for comparing video frames. But @huguesdevimeux said this won't work with fonts _because of reasons_.

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+    """
+
+    def __init__(self, text, **config):
+        logger.warning(
```
I think we should raise DeprecationWarning instead of just logging

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+        text2weight -> t2w
+        """
+        for kwargs in [config, self.CONFIG]:
+            if kwargs.__contains__("text2color"):
```
```suggestion
            if "text2color" in kwargs:
```

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+        digest_config(self, config)
+        self.original_text = text
+        text_without_tabs = text
+        if text.find("\t") != -1:
```
Question: why is it necessary to replace the tabs? What if the user wants tabs?

naveen521kk replied:
Pretty much this allow users to set custom length of tabs.  Also, tabs are added, by a default width of 4, you just try and see.

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+        self.chars = VGroup(*self.submobjects)
+        self.text = text_without_tabs.replace(" ", "").replace("\n", "")
+        nppc = self.n_points_per_cubic_curve
+        for each in self:
+            if len(each.points) == 0:
+                continue
+            points = each.points
+            last = points[0]
+            each.clear_points()
+            for index, point in enumerate(points):
+                each.append_points([point])
+                if (
+                    index != len(points) - 1
+                    and (index + 1) % nppc == 0
+                    and any(point != points[index + 1])
+                ):
+                    each.add_line_to(last)
+                    last = points[index + 1]
+            each.add_line_to(last)
```
Can you please explain what this is doing?

naveen521kk replied:
Pretty implements a way to add tabs and newlines. Copied from `Text` class.

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+            self.scale(TEXT_MOB_SCALE_FACTOR)
+
+    def remove_last_M(self, file_name: str):
+        """Internally used function. Use to format the rendered SVG files."""
```
Nitpick: there's no need to mention in the docstring that this is a function.

```suggestion
        """Internally used. Use to format the rendered SVG files."""
```

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+                self.chars[start:end].set_color_by_gradient(*gradient)
+
+    def str2style(self, string):
+        """Internally used function. Converts text to Pango Understandable Styles/"""
```
```suggestion
        """Internally used function. Converts text to Pango Understandable Styles."""
```

leotrs replied:
Question: does Pango use the word "style" to describe these? If I recall correctly, these are called "slant".

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+                self.chars[start:end].set_color(color)
+
+    def set_color_by_t2g(self, t2g=None):
+        """Internally used function.Sets gradient colours for specified strings. Same as ``set_color_by_t2c``."""
```
```suggestion
        """Internally used. Sets gradient colors for specified strings. Behaves similarly to ``set_color_by_t2c``."""
```

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+        return hasher.hexdigest()[:16]
+
+    def text2settings(self):
+        """Internally used function.
+        Convets the texts and styles to a setting for parsing
+        """
```
```suggestion
        """Internally used function. Converts the texts and styles to a setting for parsing."""
```

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +578,331 @@ def change_alignment_for_a_line(self, alignment, line_no):
+                    for start, end in self.find_indexes(word, self.text):
+                        fsw[i] = x
+                        settings.append(TextSetting(start, end, *fsw))
+        # Set All text settings(default font slant weight)
```
```suggestion
        # Set all text settings (default font, slant, weight)
```

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -717,7 +717,7 @@ def full2short(self, config):
-            if kwargs.__contains__("text2color"):
+            if "text2color" in kwargs:
                 kwargs["t2c"] = kwargs.pop("text2color")
             if kwargs.__contains__("text2font"):
```
Please change all of the others as well :)

naveen521kk replied:
will do soon.

leotrs reviewed (commented):
Not sure why the typings are necessary in the test file, but ok.

naveen521kk commented:
> Not sure why the typings are necessary in the test file, but ok.

At least, my IDE was warning me to the worse. So I had to do it. I think it doesn't hurt though.

leotrs commented:
No they don't hurt. It was just unexpected. Does your IDE bother you with all other files as well? If so, if suggest deactivating it

naveen521kk commented:
> Does your IDE bother you with all other files as well?

Pretty much. I switched to that very recently. Before I was notepad++ for my edit(sometimes Emacs as well), now I switched to VS code.

eulertour commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -575,3 +579,334 @@ def change_alignment_for_a_line(self, alignment, line_no):
+        "size": 1,
+        "slant": NORMAL,
+        "weight": NORMAL,
+        "t2c": {},
+        "t2f": {},
+        "t2g": {},
+        "t2s": {},
+        "t2w": {},
```
I would greatly prefer typing out these abbreviations to making people look them up when they see them.

naveen521kk replied:
I also wish I should. But I don't really know how to document these `config` stuff. If you can help me where to I will surely do.

naveen521kk replied:
The reason I used the same name because it is from `Text`. Pretty this was done so that this doesn't become a problem to users once they switching.

eulertour replied:
Okay, we can handle it later.

naveen521kk commented:
I think this should be merged. Also, I wouldn't be able to work on this more than this week. Also, is there anything else to do?

leotrs commented:
@naveen521kk has announced that he won't be able to spend much time on manim anymore, so I'm inclined to merge this ASAP so that we don't lose the effort he put into this. I think it's ready. Anyone else? @eulertour @behackl ?

eulertour commented:
I don't see why not if everything's working
